src/science/science.py

```diff
@@ -0,0 +1,263 @@
+                                "d1_1", "d1_2", "d1_3", "d1_4", "d1_5", "d1_6",
+                                "d2_1", "d2_2", "d2_3", "d2_4", "d2_5", "d2_6"]
+
+            # There are 18 channels.
+            # We read a uint16_t from each channel.
+            # The jetson reads byte by byte, so the program combines every two byte of information
+            # into a uint16_t.
+            count = 1
+            for var in ros_msg_variables:
+                if (not (count >= len(arr))):
+                    pass
+                    setattr(ros_msg, var, 0xFFFF & ((np.uint8(arr[count + 1]) << 8) | (np.uint8(arr[count]))))
+                else:
+                    setattr(ros_msg, var, 0)
+                count += 2
+        except Exception as e:
+            print("triad exception:", e)
+            pass
+
+    def txt_handler(self, uart_msg, ros_msg):
+        '''
+        Prints info messages revieved from nucleo to screen
+        '''
+        print(uart_msg)
+
+    def uart_send(self, uart_msg):
+        uart_msg = self.add_padding(uart_msg)
+        print(uart_msg)
+        self.ser.close()
+        self.ser.open()
+        if self.ser.isOpen():
+            self.ser.write(bytes(uart_msg, encoding='utf-8'))
+
+    async def receive(self, lcm):
+
+        # Mark TXT as always seen because they are not necessary
+        seen_tags = {tag: False if not tag == 'TXT' else True
+                     for tag in self.NMEA_HANDLE_MAPPER.keys()}
+        while True:
+            try:
+                error_counter = 0
+                tx = self.ser.readline()
+                uart_msg = str(tx)
+            except Exception as e:
+                print("Errored")
+                if error_counter < self.max_error_count:
+                    error_counter += 1
+                    print(e)
+                    await asyncio.sleep(self.sleep)
+                    continue
+                else:
+                    raise e
+            match_found = False
+            for tag, handler_func in self.NMEA_HANDLE_MAPPER.items():
+                if tag in uart_msg:  # TODO - why do we have tag in uart_msg, func is not even used
+                    print(uart_msg)
+                    match_found = True
+                    try:
+                        handler_func(uart_msg, self.NMEA_MESSAGE_MAPPER[tag])
+                        self.NMEA_PUBLISHER_MAPPER[tag].publish(self.NMEA_MESSAGE_MAPPER[tag])
+                    except Exception as e:
+                        print(e)
+                    break
+            if not match_found:
+                if not uart_msg:
+                    print('Error decoding message stream: {}'.format(uart_msg))
+            await asyncio.sleep(self.sleep)
+
+
+def main():
+
+    rospy.init_node("science")
+
+    with ScienceBridge() as bridge:
+        rospy.Subscriber("auton_led", AutonLED, bridge.auton_led_transmit)
+        rospy.Subscriber("arm_laser_cmd", Enable, bridge.mosfet_transmit, "arm_laser")
```

Comment:
if these are  commands that are meant to be used as RPCs as opposed to continious datastreams it might be better to implement them as services http://wiki.ros.org/Services

Reply:
Oh dang, that actually looks super useful, this could really clean up a lot of the communication between the GUIs and ESW bridge programs.

Reply:
Ok I'm reading up on services and they look pretty good.
The only concern is that only one program can open a serial port at a time, so services probably don't make sense. If someone tried to force the implementation of services, they would need to constantly reopen and close the serial port on the src/science program (which is not feasible since src/science will constantly need to read in every line of data).

LMK if you have any thoughts on this.

Reply:
Ok I thought about it again, and it might be possible to reopen and close the serial port, since we only open it for a split second to read the line. Again, I'm still not sure if this is guaranteed to work

Reply:
As a follow up, ESW constantly posts data (such as spectral data and thermistor data), but should this be changed to a service too? Online, it says that topics should be used for continuous data streams, but we only post this data once every second.

Reply:
the driver can own the serial port, the service calls are just to request read and write data. For the constant data posts you should stick to pub/sub